pybboxes/types/albumentations_bounding_box.py

```diff
@@ -34,6 +34,28 @@ def to_voc(self, return_values: bool = False) -> Union[Tuple[int, int, int, int]
             return x_tl, y_tl, x_br, y_br
         return BoundingBox(x_tl, y_tl, x_br, y_br, image_size=self.image_size, strict=self.strict)
 
+    def shift(self, horizontal_threshold: float, vertical_threshold: float) -> "AlbumentationsBoundingBox":
```

Comment:
Please, take this comment as for other methods as well. 

(1) Since we convert to VOC, the type hints as well as the docstring needs to be changed, only integers are accepted since we use VOC type under the hood for shifting.
(2) We need to clearly desribe and explicitly write in the docstring that shifting is performed in pixel values not in normalized values.

Reply:
I did not understand what do you mean by "pixel values not in normalised values". Could you clarify a little more ?

Reply:
Sure. I meant unnormalized values, like those shift values (`horizontal_threshold`, `vertical_threshold`) must be always in pixel values integers because VOC style is in pixel values (x,y coordinates). For example YOLO style is `(x-center, y-center, width, height)` but these values are normalized (w.r.t image_size) values and not pixel coordinates.

Currently, in the implementation in this PR, regardless of the box style the box is first converted to VOC style, and addition is then applied. That's why, for example, we cannot give normalized shift values to YOLO style values (normalized). Alternatively, if we take the input shift values in that particular box style, then VOC style conversion must be omitted. Further, see the pseudo snippet below

```python
yolo_values = [0.4046875, 0.840625, 0.503125, 0.24375]
yolo_bbox = BoundingBox.from_yolo(*yolo_values, image_size=(640,480))

# I want to shift my YOLO box
shifted_yolo_box = yolo_bbox.shift(0.07, 0.02)
```

Now, when I call `.shift()` internally these steps happen; (1) yolo_values are converted to voc values, (2) my shift values are added to the associated values `0.07`, `0.02` respectively.

At step (2) yolo shift values `0.07, 0.02` are added to VOC values which are `[98, 345, 420, 462]` which is not desired. There are two solutions that comes up,

### Solution 1 (Current)
Accept shift values as unnormalized. Convert the box values to voc, add the shift values, and return.

### Solution 2
Omit converting values to VOC, directly apply the shift values and then instantiate/construct the object and return.


Both solutions are fine, and actually I guided you towards solution 1 although currently when I consider the case again, solution 2 seems a better alternative. In this case, I kinda misguided you towards solution 1. Thus, I see currently two options,

- Keep solution 1 (current implementation), but fix the accept type (all should be int because of converting VOC) and also add an info to the docstring about shift values must be unnormalized.
- Switch to solution 2 (might be better).

@harshraj22 wdyt of these alternatives ?

Reply:
Yes, I too would prefer solution 2. So if I understand correctly, it would be overwriting `shift` method in each BoundingBox type.

Reply:
> Yes, I too would prefer solution 2. So if I understand correctly, it would be overwriting `shift` method in each BoundingBox type.

Your implementation so far will remain as is except the to_voc conversion at step 1 since we'd like to update the values associated with each style. In short, simply remove to_voc parts in your implementation and all is fine.

Reply:
<img width="1257" alt="image" src="https://user-images.githubusercontent.com/46635452/173391908-d55660c9-0d0d-4806-a73e-47b4e1129627.png">

Sounds like there's some conversion logic required with the `image` as well in all these types.

Reply:
> Sounds like there's some conversion logic required with the `image` as well in all these types.

Well, in case of out of box errors that's fine and the method does not need to bother with validating the values. However, here a strange thing is happening because `(w, h)` are not changed, but it gives an error, very odd indeed. It may be due to a bug. I'll try to reproduce this problem on your fork and I'll write my findings here, and we can proceed afterwards.

Reply:
Sure. Something sounds fishy. I have pushed my local changes in the latest commit so that it is easy for you to reproduce.

Reply:
Well, in your fork for YOLO box everything seems fine, with the code snippet below, I got the desired output:

```python
>>> yolo_values = [0.4046875, 0.840625, 0.503125, 0.24375]
>>> yolo_bbox = YoloBoundingBox.from_array(yolo_values, image_size=(640, 480))
>>> shifted_yolo_bbox = yolo_bbox.shift((0.08, 0.03))
>>> print(yolo_bbox)
>>> print(shifted_yolo_bbox)
<[0.4047 0.8406 0.5031 0.2437] (322x117) | Image: (640x480)>
<[0.4847 0.8706 0.5031 0.2437] (322x117) | Image: (640x480)>
```
,and this

```python
>>> oob_bbox = yolo_bbox.shift((0.08, 0.2))
>>> print(oob_bbox)
Traceback (most recent call last):
  File "C:\Users\devri\lab\projects\pybboxes\pybboxes\types\yolo_bounding_box.py", line 102, in <module>
    oob_bbox = yolo_bbox.shift((0.08, 0.2))
  File "C:\Users\devri\lab\projects\pybboxes\pybboxes\types\yolo_bounding_box.py", line 61, in shift
    shifted_bbox = YoloBoundingBox(
  File "C:\Users\devri\lab\projects\pybboxes\pybboxes\types\yolo_bounding_box.py", line 17, in __init__
    super(YoloBoundingBox, self).__init__(x_c, y_c, w, h, image_size=image_size, strict=strict)
  File "C:\Users\devri\lab\projects\pybboxes\pybboxes\types\base.py", line 67, in __init__
    self._validate_values(v1, v2, v3, v4)
  File "C:\Users\devri\lab\projects\pybboxes\pybboxes\types\yolo_bounding_box.py", line 24, in _validate_values
    raise ValueError("Given bounding box values is out of bounds.")
ValueError: Given bounding box values is out of bounds.
```

All in all, everything seems fine. Can you give minimal example to reproduce your case ? Can you accidentally shift the box out of bounds ? @harshraj22

Reply:
That's strange ! macOS Monterey, python3.8.9
to reproduce, from the latest commit on my branch, run `pytest` from the pytest directory.

Reply:
Hi @harshraj22, sorry for late reply. Well, I've checked it. The tests failing simply because the shift amount tries to shift the boxes out of bounds,

```python
# Albumentations box
<[0.1531 0.7188 0.6562 0.9625] (322x117) | Image: (640x480)>

# the shift amount
(2, 2)
```

Albumentations, fiftyone and YOLO are bounding box styles with normalized values meaning that the bounding boxes that are not OOB needs to be in (0,1) range, clearly with the shift `(2,2)` they are pushed towards OOB. My advise is that, define two distinct fixtures for normalized shift, and unnormalized shift and apply accordingly.

Also, test case for COCO is failing due to mismatch between `actual` being list, and `desired` being tuple, simply just match the types either make both lists or both tuples, make them consistent among all test cases so chose one type for all test cases.

Reply:
The `shift( )` method takes `Tuple[int, int]` as arg right ? We can't pass a tuple of values in (0, 1) ?

Reply:
> The `shift( )` method takes `Tuple[int, int]` as arg right ? We can't pass a tuple of values in (0, 1) ?

No, we've picked the second option (see [Solution 2](https://github.com/devrimcavusoglu/pybboxes/pull/9#discussion_r895499391)), and thus on `shift()` method every box takes their associated values. That is, normalized box styles (Albumentations, Fiftyone, YOLO) will take floats between `(0,1)`, and unnormalized box styles (COCO, VOC) will take integers.